File: implementations/elixir/lib/rhizomatic/delta.ex
defmodule Rhizomatic.Delta do
  @moduledoc """
  The delta (SPEC-1): claims validation at the boundary (reject, never
  repair), canonical serialization (§4.1), and content addressing.

  In-memory claims form (all boundary-validated):

      %{
        timestamp: float(),           # finite f64; a native integer is REJECTED (D14)
        author: String.t(),           # non-empty, NFC
        pointers: [pointer]           # at least one
      }

      pointer :: %{role: String.t(), target: target}
      target ::
          {:string, String.t()}
        | {:number, float()}
        | {:boolean, boolean()}
        | {:entity, id :: String.t(), context :: String.t() | nil}
        | {:delta, hex :: String.t(), context :: String.t() | nil}
        | {:bytes, mime :: String.t(), payload :: binary()}

  The host-boundary numeric policy (SPEC-1 §4.1 / ERRATA D14): on the BEAM,
  `42` and `42.0` are distinct terms, so claim construction rejects native
  integers everywhere a number is expected. The one blessed coercion point is
  the JSON debug profile parser (`Rhizomatic.Profile`), where an integer
  token is unambiguously a float spelling.
  """

  alias Rhizomatic.{Cbor, Hash}

  @type target ::
          {:string, String.t()}
          | {:number, float()}
          | {:boolean, boolean()}
          | {:entity, String.t(), String.t() | nil}
          | {:delta, String.t(), String.t() | nil}
          | {:bytes, String.t(), binary()}

  @type pointer :: %{role: String.t(), target: target()}
  @type claims :: %{timestamp: float(), author: String.t(), pointers: [pointer()]}

  # ------------------------------------------------------------- validation

  @doc """
  Validate claims at the boundary. Returns `{:ok, claims}` or
  `{:error, reason}`. Reject, never repair (SPEC-4 §2).
  """
  @spec validate(term()) :: {:ok, claims()} | {:error, term()}
  def validate(%{timestamp: ts, author: author, pointers: pointers} = claims)
      when map_size(claims) == 3 do
    with :ok <- validate_timestamp(ts),
         :ok <- validate_nonempty_nfc(author, :author),
         :ok <- validate_pointers(pointers) do
      {:ok, %{timestamp: ts, author: author, pointers: pointers}}
    end
  end

  def validate(_), do: {:error, :malformed_claims}

  defp validate_timestamp(ts) when is_float(ts) do
    # NaN/Infinity are not representable as Elixir floats; finiteness holds
    :ok
  end

  defp validate_timestamp(ts) when is_integer(ts), do: {:error, {:native_integer, :timestamp}}
  defp validate_timestamp(_), do: {:error, :timestamp_not_a_number}

  defp validate_pointers(pointers) when is_list(pointers) and pointers != [] do
    Enum.reduce_while(pointers, :ok, fn p, :ok ->
      case validate_pointer(p) do
        :ok -> {:cont, :ok}
        {:error, e} -> {:halt, {:error, e}}
      end
    end)
  end

  defp validate_pointers([]), do: {:error, :pointers_empty}
  defp validate_pointers(_), do: {:error, :pointers_not_a_list}

  defp validate_pointer(%{role: role, target: target} = p) when map_size(p) == 2 do
    with :ok <- validate_nonempty_nfc(role, :role), do: validate_target(target)
  end

  defp validate_pointer(_), do: {:error, :malformed_pointer}

  defp validate_target({:string, s}), do: validate_nfc(s, :string_primitive)
  defp validate_target({:number, f}) when is_float(f), do: :ok
  defp validate_target({:number, i}) when is_integer(i), do: {:error, {:native_integer, :target}}
  defp validate_target({:boolean, b}) when is_boolean(b), do: :ok

  defp validate_target({:entity, id, context}) do
    with :ok <- validate_nonempty_nfc(id, :entity_id), do: validate_context(context)
  end

  defp validate_target({:delta, hex, context}) do
    with :ok <- validate_nonempty_nfc(hex, :delta_ref), do: validate_context(context)
  end

  defp validate_target({:bytes, mime, payload}) do
    cond do
      not is_binary(payload) -> {:error, :bytes_value_not_binary}
      true -> validate_nonempty_nfc(mime, :mime)
    end
  end

  defp validate_target(_), do: {:error, :malformed_target}

  defp validate_context(nil), do: :ok
  defp validate_context(c), do: validate_nonempty_nfc(c, :context)

  defp validate_nonempty_nfc(s, field) do
    cond do
      not is_binary(s) -> {:error, {:not_a_string, field}}
      s == "" -> {:error, {:empty_string, field}}
      true -> validate_nfc(s, field)
    end
  end

  # NFC is validated at the boundary, never repaired (SPEC-1 §4.1 / D11)
  defp validate_nfc(s, field) do
    cond do
      not is_binary(s) or not String.valid?(s) -> {:error, {:not_a_string, field}}
      :unicode.characters_to_nfc_binary(s) != s -> {:error, {:not_nfc, field}}
      true -> :ok
    end
  end

  # --------------------------------------------------------- canonical form

  @doc "Canonical deterministic-CBOR bytes of validated claims (SPEC-1 §4.1)."
  @spec canonical_bytes(claims()) :: binary()
  def canonical_bytes(claims) do
    Cbor.encode(to_cbor(claims))
  end

  @doc "Lowercase-hex multihash id of validated claims."
  @spec id_hex(claims()) :: String.t()
  def id_hex(claims), do: Hash.id_hex(canonical_bytes(claims))

  @doc "34 raw multihash bytes of the id (the Ed25519 signing message, §5)."
  @spec id_bytes(claims()) :: binary()
  def id_bytes(claims), do: Hash.content_address(canonical_bytes(claims))

  @doc false
  @spec to_cbor(claims()) :: Cbor.t()
  def to_cbor(%{timestamp: ts, author: author, pointers: pointers}) do
    {:map,
     [
       {{:tstr, "author"}, {:tstr, author}},
       {{:tstr, "pointers"}, {:arr, Enum.map(pointers, &pointer_to_cbor/1)}},
       {{:tstr, "timestamp"}, {:float, ts}}
     ]}
  end

  defp pointer_to_cbor(%{role: role, target: target}) do
    {:map,
     [
       {{:tstr, "role"}, {:tstr, role}},
       {{:tstr, "target"}, target_to_cbor(target)}
     ]}
  end

  defp target_to_cbor({:string, s}), do: {:tstr, s}
  defp target_to_cbor({:number, f}), do: {:float, f}
  defp target_to_cbor({:boolean, b}), do: {:bool, b}

  defp target_to_cbor({:entity, id, context}),
    do: {:map, [{{:tstr, "id"}, {:tstr, id}} | context_pair(context)]}

  defp target_to_cbor({:delta, hex, context}),
    do: {:map, [{{:tstr, "delta"}, {:tstr, hex}} | context_pair(context)]}

  defp target_to_cbor({:bytes, mime, payload}),
    do: {:map, [{{:tstr, "mime"}, {:tstr, mime}}, {{:tstr, "value"}, {:bstr, payload}}]}

  defp context_pair(nil), do: []
  defp context_pair(c), do: [{{:tstr, "context"}, {:tstr, c}}]
end

File: implementations/haskell/src/Rhizomatic/Delta.hs
{-# LANGUAGE OverloadedStrings #-}

-- | The delta layer (SPEC-1 §2 / §4): typed claims, boundary validation,
-- canonical bytes, content address.
--
-- D14 note (host-boundary numeric policy): in this witness the hazard is
-- discharged by the types — a claims number is only ever a 'Double'; there is
-- no constructor accepting 'Integer', so no native integer term can reach
-- claim construction to be coerced. The one blessed coercion point remains
-- the JSON debug profile parser ('Rhizomatic.Profile').
module Rhizomatic.Delta
  ( Target (..),
    Pointer (..),
    Claims (..),
    validateClaims,
    claimsToCbor,
    canonicalBytes,
    deltaId,
    deltaIdHex,
  )
where

import qualified Data.ByteString as B
import qualified Data.Text as T
import Rhizomatic.Blake3 (blake3)
import Rhizomatic.Cbor (Item (..), encode)
import Rhizomatic.Hex (encodeHex)

data Target
  = TEntity T.Text (Maybe T.Text) -- id, context?
  | TDeltaRef T.Text (Maybe T.Text) -- delta hex, context?
  | TString T.Text
  | TNumber Double
  | TBool Bool
  | TBytes T.Text B.ByteString -- mime, raw payload
  deriving (Eq, Show)

data Pointer = Pointer {pRole :: T.Text, pTarget :: Target} deriving (Eq, Show)

data Claims = Claims {cTimestamp :: Double, cAuthor :: T.Text, cPointers :: [Pointer]} deriving (Eq, Show)

-- | SPEC-1 §2.1 boundary validation: reject, never repair.
validateClaims :: Claims -> Either String ()
validateClaims (Claims ts author ptrs) = do
  finiteNumber "timestamp" ts
  nonEmpty "author" author
  if null ptrs then Left "a delta MUST contain at least one pointer" else Right ()
  mapM_ validatePointer ptrs

validatePointer :: Pointer -> Either String ()
validatePointer (Pointer role target) = do
  nonEmpty "role" role
  case target of
    TEntity _ ctx -> maybe (Right ()) (nonEmpty "context") ctx
    TDeltaRef _ ctx -> maybe (Right ()) (nonEmpty "context") ctx
    TString _ -> Right ()
    TNumber n -> finiteNumber "number primitive" n
    TBool _ -> Right ()
    TBytes mime _ -> nonEmpty "mime" mime

finiteNumber :: String -> Double -> Either String ()
finiteNumber what d
  | isNaN d || isInfinite d = Left (what ++ " must be a finite number")
  | otherwise = Right ()

-- D16: strings are byte-honest — any valid UTF-8 is admitted; NFC is authoring hygiene
-- (SPEC-5 §6), not a boundary law. Only emptiness is checked where the spec demands it.
nonEmpty :: String -> T.Text -> Either String ()
nonEmpty what t
  | T.null t = Left (what ++ " must be non-empty")
  | otherwise = Right ()

-- | SPEC-1 §4.1 claims layout. Callers validate first; this is pure layout.
claimsToCbor :: Claims -> Item
claimsToCbor (Claims ts author ptrs) =
  Map
    [ ("author", TStr author),
      ("pointers", Arr (map pointerToCbor ptrs)),
      ("timestamp", Num ts)
    ]

pointerToCbor :: Pointer -> Item
pointerToCbor (Pointer role target) = Map [("role", TStr role), ("target", targetToCbor target)]

targetToCbor :: Target -> Item
targetToCbor t = case t of
  TEntity eid ctx -> Map (("id", TStr eid) : ctxKv ctx)
  TDeltaRef d ctx -> Map (("delta", TStr d) : ctxKv ctx)
  TString s -> TStr s
  TNumber n -> Num n
  TBool b -> Bool' b
  TBytes mime payload -> Map [("mime", TStr mime), ("value", BStr payload)]
  where
    ctxKv = maybe [] (\c -> [("context", TStr c)])

canonicalBytes :: Claims -> Either String B.ByteString
canonicalBytes claims = do
  validateClaims claims
  Right (encode (claimsToCbor claims))

-- | 34-byte multihash: 0x1e (blake3 multicodec) + 0x20 (length) + digest.
deltaId :: Claims -> Either String B.ByteString
deltaId claims = do
  bytes <- canonicalBytes claims
  Right (B.pack [0x1e, 0x20] <> blake3 bytes)

deltaIdHex :: Claims -> Either String String
deltaIdHex claims = encodeHex <$> deltaId claims
